File: leet/0131_palindrome_partitioning/0131_palindrome_partitioning.js
// 131. Palindrome Partitioning
// https://leetcode.com/problems/palindrome-partitioning/
// O(n*2^n) time, O(n^2) space
var partition = function(s) {
    const solutions = [];
    dp = Array(s.length);
    const isPali = (s, i, j) => {
        for (; i < j; i++, j--)
            if (s[i] !== s[j])
                return false;
        return true;
    }
    const makeDP = () => {
        for (let i = 0; i < s.length; i++) {
            dp[i] = [];
            for (let j = 0; i + j < s.length; j++) {
                if (isPali(s, i, i + j))
                    dp[i].push(s.slice(i, i + j + 1));
            }
        }
    }
    makeDP();
    
    const dfs = (i, partition) => {
        if (i === s.length) {
            solutions.push([...partition])
            return;
        }
        for (const pali of dp[i]) {
            partition.push(pali);
            dfs(i + pali.length, partition);
            partition.pop();
        }
    }
    dfs(0, []);
    return solutions;
};

File: leet/0131_palindrome_partitioning/0131_palindrome_partitioning.cpp
// LeetCode 131. Palindrome Partitioning
// https://leetcode.com/problems/palindrome-partitioning/
// O(n*2^n) time-and-space, bottom-up DP + backtracking
class Solution {
public:
    vector<vector<string>> dp;
    vector<vector<string>> partitions;
    
    vector<vector<string>> partition(string s) {
        makeDP(s);
        vector<string> partition;
        dfs(s, 0, partition);
        return partitions;
    }
    
    private: void makeDP(string &s) {
        dp.resize(s.size());
        for (size_t i = 0; i < s.length(); i++) {
            for (size_t j = i; j < s.length(); j++) {
                if (isPali(s, i, j + 1))
                    dp[i].push_back(s.substr(i, j - i + 1));
            }
        }
    }
    
    private: bool isPali(string &s, size_t i, size_t j) {
        while (i < j)
            if (s[i++] != s[--j])
                return false;
        return true;
    }
    
    private: void dfs(string &s, size_t i, vector<string> &partition) {
        if (i == s.size()) {
            partitions.push_back(partition);
            return;
        }
        for (auto pali : dp[i]) {
            partition.push_back(pali);
            dfs(s, i + pali.size(), partition);
            partition.pop_back();
        }
    }
};
